R/datashield.sessions.R

```diff
@@ -0,0 +1,133 @@
+#' R/DataSHIELD remote sessions
+#'
+#' Ensure that the remote R sessions are up and running during the analysis.
+#'
+#' @param conns \code{\link{DSConnection-class}} object or a list of \code{\link{DSConnection-class}}s.
+#' @param async Whether the remote R/DataSHIELD session should be created asynchronously. When TRUE (default) the calls are parallelized over
+#'   the connections, when the connection supports that feature, with an extra overhead of requests.
+#' @param success Callback function that will be called each time an R session has been created from a connection. 
+#'   The expected function signature is the connection/study name. Default is NULL (no callback).
+#' @param error Callback function that will be called each time the R session creation request has failed. 
+#'   The expected function signature is the connection/study name and the error message. Default is NULL (no callback).
+#' @param errors.print Boolean, whether to print datashield errors in the console or return a message indicating that they can be retrieved using `datashield.errors`.
+#'
+#' @examples
+#'\dontrun{
+#' # call sessions function on server side asynchronously
+#' # i.e. each study connection will create a remote R session in parallel
+#' datashield.sessions(conns)
+#' 
+#' # call sessions function with callback functions
+#' result <- datashield.sessions(conns,
+#'   success = function(server) {
+#'     # do something with server's success
+#'   },
+#'   error = function(server, error) {
+#'     # do something with server's error
+#'   })
+#' }
+#'
+#' @export
+datashield.sessions <- function(conns, async=TRUE, success=NULL, error=NULL, errors.print = getOption("datashield.errors.print", FALSE)) {
+  .clearLastErrors()
+
+  if (is.list(conns)) {
+    # filter conns supporting session API and not having connection
+    fconns <- Filter(function(conn) { !is.null(dsIsAsync(conn)$session) && !dsHasSession(conn) }, conns)
+    if (length(fconns) == 0) {
+      return(invisible(NULL))
+    }
+    
+    sessions <- list()
+    asyncs <- lapply(fconns, function(conn) { ifelse(async, dsIsAsync(conn)$session, FALSE) })
+    pb <- .newProgress(total = 1 + length(fconns))
+    # async first
+    for (n in names(fconns)) {
+      if(asyncs[[n]]) {
+        tryCatch({
+          sessions[[n]] <- dsSession(fconns[[n]], async=async)
+        }, error = function(e) {
+          .appendError(n, conditionMessage(e))
```

Comment:
This is repeated - could you break into a separate helper function which is then called in the error part of the trycatch?